Experiments based on SpecAugment change

luomingshuang opened the pull request:
As a response for #201 ,  in the new asr_datamodule.py, I set `num_frame_masks` to 10 and `max_frames_mask_fraction` as 0.4 according to Dan's suggestions. Also, I plan to use conformer_ctc as the baseline. For decreasing the training time, I plan to verify this idea based on test-clean 100h data for training.

danpovey commented on egs/librispeech/ASR/tdnn_lstm_ctc/asr_datamodule.py:
```diff
@@ -383,3 +383,212 @@ def test_clean_cuts(self) -> CutSet:
+        features_mask_size: int = 13,
+        num_frame_masks: int = 1,
+        frames_mask_size: int = 70,
+        max_frames_mask_fraction: float = 0.2,
```
You can probably change the defaults here too, since the behavior/meaning of those options has changed.
And make sure to correctly document the new meaning of the options, e.g.:
clarify frames_mask_size is per masked region, num_frame_masks is the maximum (it's also constrained by max_frames_mask_fraction), and that max_frames_mask_fraction is the total masked fraction including all masked regions (the expected/mean fraction will be less than half of this, as we randomly select the mask sizes between 0 and the computed maximum mask lengths).

danpovey commented:
You may be able to use a Jupyter notebook to test this and display the features, Lhotse has good interoperation with Jupyter (see if the lhotse page has any links to such notebooks).

luomingshuang commented:
OK.
> You may be able to use a Jupyter notebook to test this and display the features, Lhotse has good interoperation with Jupyter (see if the lhotse page has any links to such notebooks).

luomingshuang commented:
I test a wav sample from librispeech and display the feature with the colab url (https://colab.research.google.com/drive/1mwZJDkub5fRSTTQC53hXBQJrvTXBwMBG?usp=sharing). 

The heatmap of the original fbank feature shows as follows:
![image](https://user-images.githubusercontent.com/37799481/152950333-08d2cb40-09bf-4eb6-8b0a-27327b223e2d.png)


The heatmap of the old specaugment version (num_frame_masks: int = 1, max_frames_mask_fraction: float = 0.2, also with 
```
max_mask_frames = min(
                    self.frames_mask_size,
                    self.max_frames_mask_fraction * features.size(0),
                )
```
shows as follows:
![image](https://user-images.githubusercontent.com/37799481/152950706-2df4ea3d-7ae2-44cd-9957-fc42105a632e.png)


The heatmap of the modified specaugment version (num_frame_masks: int = 10, max_frames_mask_fraction: float = 0.4, also with 
```
_max_tot_mask_frames = self.max_frames_mask_fraction * features.size(0)
num_frame_masks = min(self.num_frame_masks, math.ceil(_max_tot_mask_frames / self.frames_mask_size))
max_mask_frames = min(self.frames_mask_size, _max_tot_mask_frames // num_frame_masks)
```
shows as folows:
![image](https://user-images.githubusercontent.com/37799481/152952046-ae594122-c814-4a1e-8cae-051534e8e88e.png)

danpovey commented:
OK.  I think perhaps you can reduce max_frames_mask_fraction to 0.3; 0.4 seems to give too much masking.
BTW, that implementation does not scale efficiently for large num_frame_masks, because it copies the entire 
matrix for each mask.  It would be better to accumulate a boolean 1-d tensor and apply it as a mask later.
That would require reorganizing the code a little.
Perhaps you could work on making it more efficient, since the GPUs are very busy right now anyway?

luomingshuang commented:
Em, I see. In the original version, the self.num_frame_masks is 1, so it won't influence the running efficiency. When increasing the num_frame_masks, it will influence the efficiency. 
It seems that the `for` loop is not necessary. Can I remove the  `for` loop? And change 
```
            for _ in range(self.num_feature_masks):
                features = mask_along_axis(
                    features.unsqueeze(0),
                    mask_param=self.features_mask_size,
                    mask_value=mean,
                    axis=2,
                ).squeeze(0)
            for _ in range(self.num_frame_masks):
                _max_tot_mask_frames = self.max_frames_mask_fraction * features.size(0)
                num_frame_masks = min(self.num_frame_masks, math.ceil(_max_tot_mask_frames / self.frames_mask_size))
                max_mask_frames = min(self.frames_mask_size, _max_tot_mask_frames // num_frame_masks)
                
                features = mask_along_axis(
                    features.unsqueeze(0),
                    mask_param=max_mask_frames,
                    mask_value=mean,
                    axis=1,
                ).squeeze(0)
```
to 
```
            features = mask_along_axis(
                    features.unsqueeze(0),
                    mask_param=self.features_mask_size,
                    mask_value=mean,
                    axis=2,
            ).squeeze(0)
           
            _max_tot_mask_frames = self.max_frames_mask_fraction * features.size(0)
            num_frame_masks = min(self.num_frame_masks, math.ceil(_max_tot_mask_frames / self.frames_mask_size))
            max_mask_frames = min(self.frames_mask_size, _max_tot_mask_frames // num_frame_masks)
                
            features = mask_along_axis(
                    features.unsqueeze(0),
                    mask_param=max_mask_frames,
                    mask_value=mean,
                    axis=1,
             ).squeeze(0)
```
?
> OK. I think perhaps you can reduce max_frames_mask_fraction to 0.3; 0.4 seems to give too much masking. BTW, that implementation does not scale efficiently for large num_frame_masks, because it copies the entire matrix for each mask. It would be better to accumulate a boolean 1-d tensor and apply it as a mask later. That would require reorganizing the code a little. Perhaps you could work on making it more efficient, since the GPUs are very busy right now anyway?

luomingshuang commented:
Oh, it seems the `for` loop can not remove.
> Em, I see. In the original version, the self.num_frame_masks is 1, so it won't influence the running efficiency. When increasing the num_frame_masks, it will influence the efficiency. It seems that the `for` loop is not necessary. Can I remove the `for` loop? And change
> 
> ```
>             for _ in range(self.num_feature_masks):
>                 features = mask_along_axis(
>                     features.unsqueeze(0),
>                     mask_param=self.features_mask_size,
>                     mask_value=mean,
>                     axis=2,
>                 ).squeeze(0)
>             for _ in range(self.num_frame_masks):
>                 _max_tot_mask_frames = self.max_frames_mask_fraction * features.size(0)
>                 num_frame_masks = min(self.num_frame_masks, math.ceil(_max_tot_mask_frames / self.frames_mask_size))
>                 max_mask_frames = min(self.frames_mask_size, _max_tot_mask_frames // num_frame_masks)
>                 
>                 features = mask_along_axis(
>                     features.unsqueeze(0),
>                     mask_param=max_mask_frames,
>                     mask_value=mean,
>                     axis=1,
>                 ).squeeze(0)
> ```
> 
> to
> 
> ```
>             features = mask_along_axis(
>                     features.unsqueeze(0),
>                     mask_param=self.features_mask_size,
>                     mask_value=mean,
>                     axis=2,
>             ).squeeze(0)
>            
>             _max_tot_mask_frames = self.max_frames_mask_fraction * features.size(0)
>             num_frame_masks = min(self.num_frame_masks, math.ceil(_max_tot_mask_frames / self.frames_mask_size))
>             max_mask_frames = min(self.frames_mask_size, _max_tot_mask_frames // num_frame_masks)
>                 
>             features = mask_along_axis(
>                     features.unsqueeze(0),
>                     mask_param=max_mask_frames,
>                     mask_value=mean,
>                     axis=1,
>              ).squeeze(0)
> ```
> 
> ?
> 
> > OK. I think perhaps you can reduce max_frames_mask_fraction to 0.3; 0.4 seems to give too much masking. BTW, that implementation does not scale efficiently for large num_frame_masks, because it copies the entire matrix for each mask. It would be better to accumulate a boolean 1-d tensor and apply it as a mask later. That would require reorganizing the code a little. Perhaps you could work on making it more efficient, since the GPUs are very busy right now anyway?

danpovey commented:
It would be necessary to reorganize the code.  The for loop could make modifications to a 1-d bool tensor.
it's OK to do experiments with the current code though, it may not really affect efficiency much.

luomingshuang commented:
Ok, I am trying it.
> It would be necessary to reorganize the code. The for loop could make modifications to a 1-d bool tensor. it's OK to do experiments with the current code though, it may not really affect efficiency much.

luomingshuang commented:
It seems that the modified specaugment has a positive effect on the performance. I run the conformer-ctc based on the modified specaugment as this pr. 

At present, there is a comparison between the modified version and the original version when both trained for 15 epochs:
```
-----------------------------------------------
ctc-decoding, epoch=14, avg=1
-----------------------------------------------
      method\dataset |  test-clean(WER%) | test-other(WER%)
-----------------------------------------------
modified specaugment |     14.05         |  32.47
-----------------------------------------------
original specaugment |     15.15         |  33.76
-----------------------------------------------
```
Will continue train some epochs and have a see forward.

danpovey commented:
Cool!
BTW, I tested after 10 epochs using a librispeech-100h setup I have, training with 1 GPU.
It's a nonstandard setup, with relu on the attention and warm-steps=8000.
It was slightly worse after 10 epochs (--epoch 9 --avg 3), with WER (9.3,24.6 -> 9.85, 25.3), 
but the validation probs look better.  I will test again after more epochs, it could be it makes it
worse for the first few epochs.
I was using 0.3 for max_mask_fraction.

danpovey commented:
I am also, separately, testing the effect of changing p=0.5 to p=0.9 while using the original SpecAugment code.  It is only on epoch 0, I'll let you know when I have some results.

luomingshuang commented:
OK, and the above my results are based on max_mask_fraction=0.6.

pzelasko commented:
Cool! If you can report the best configuration and implementation at the end of this excercise, we could update Lhotse to reflect those.

danpovey commented:
I updated my comment above
https://github.com/k2-fsa/icefall/pull/205#issuecomment-1033394646
... just the baseline SpecAug but with p=0.5 changed to p=0.9 gave even more improvement.
I'm working on a combination but as I mentioned, I'm having problems with convergence right now
(the particular setup I'm using is tricky).

luomingshuang commented:
Yes, in the modified specaugment, when I use num_frame_masks=10, max_frames_mask_fraction=0.4, p=0.8, the model can't converge. But when p=0.5, the model can converge.

luomingshuang commented:
I am doing a experiment with the original specaugment but with num_frame_masks=10, max_frames_mask_fraction=0.4 and p=0.5. I want to see if there is an improvement when modifying the two parameters for the original specaugment.

danpovey commented:
I don't think those settings make much sense for the original SpecAugment.  num_frame_masks=10 will completely mask the signal, for signals less than about 5 seconds or so long.  And max_frames_mask_fraction is interpreted per masked region, so it will never have an effect unless the signal is less than 2.5 seconds long.

danpovey commented on egs/librispeech/ASR/tdnn_lstm_ctc/asr_datamodule.py:
```diff
@@ -383,3 +383,212 @@ def test_clean_cuts(self) -> CutSet:
+            for _ in range(self.num_frame_masks):
+                _max_tot_mask_frames = self.max_frames_mask_fraction * features.size(0)
+                num_frame_masks = min(self.num_frame_masks, math.ceil(_max_tot_mask_frames / self.frames_mask_size))
+                max_mask_frames = min(self.frames_mask_size, _max_tot_mask_frames // num_frame_masks)
```
@luomingshuang this code is not right.  You are setting num_frame_masks here but not doing anything with it: the code iterates by self.num_frame_masks.

luomingshuang replied:
Em, you mean I should do changes as follows:
```
_max_tot_mask_frames = self.max_frames_mask_fraction * features.size(0)
num_frame_masks = min(self.num_frame_masks, math.ceil(_max_tot_mask_frames / self.frames_mask_size))
max_mask_frames = min(self.frames_mask_size, _max_tot_mask_frames // num_frame_masks)

for _ in range(num_frame_masks):
                features = mask_along_axis(
                    features.unsqueeze(0),
                    mask_param=max_mask_frames,
                    mask_value=mean,
                    axis=1,
                ).squeeze(0)
```

danpovey commented on egs/librispeech/ASR/conformer_ctc/asr_datamodule.py:
```diff
@@ -1 +1,594 @@
-../tdnn_lstm_ctc/asr_datamodule.py
\ No newline at end of file
+# Copyright      2021  Piotr Żelasko
```
I think you changed this file from a soft link to a file.

luomingshuang replied:
Yes. I changed it to a file.

csukuangfj commented on egs/librispeech/ASR/conformer_ctc/asr_datamodule.py:
```diff
@@ -1 +1,594 @@
+        return load_manifest(self.args.manifest_dir / "cuts_test-other.json.gz")
+
+
+import math
```
Why not to move it to a separate file and then import it ?

luomingshuang replied:
Of course. A separate file is also a good choose. Will notice it.

danpovey replied:
This may create issues with needing to have too many soft links, if the file is in the recipes dir.
Eventually  Piotr plans to take our changes into Lhotse, that may be the easiest way...

luomingshuang commented:
After correcting the SpecAugment, the result has a significant progress comparing with the former or the the baseline. Will continue to do experiments based on this correct specaugment.
```
-----------------------------------------------------------------------
                  ctc-decoding, epoch=14, avg=1
-----------------------------------------------------------------------
        method\dataset       |  test-clean(WER%) | test-other(WER%)
-----------------------------------------------------------------------
modified specaugment-correct |     13.10         |  31.49
-----------------------------------------------------------------------
modified specaugment-wrong   |     14.05         |  32.47
-----------------------------------------------------------------------
     original specaugment    |     15.15         |  33.76
-----------------------------------------------------------------------
```

danpovey commented:
Great progress!
We can expect a little less improvement on the full Librispeech I think, but still some...

danpovey commented:
@luomingshuang I have updated my table above, again; I now have results with max_mask_fraction=0.3, p=0.9, but disappointingly, it is worse than the baseline code with p=0.9.  I am trying a couple other configurations.

danpovey commented:
Yes, I think 0.4 is too big.  You could try 0.2, see if that converges, maybe you will be able to replicate my result.

danpovey commented:
I have added more results to the table above.  So far my best results are with mask_fraction=0.15, p=0.9; I got 8.05,20.92, versus our original SpecAug implementation with p=0.5: 9.3/24.55

pzelasko commented:
Did you ever try p=1.0? I think in the original paper (https://arxiv.org/pdf/1904.08779v1.pdf) they never specified the probability of transform (`p` in the paper has - quite misleading - meaning of `mask_fraction` in this discussion), which to me implies they just applied the transform on all data.

danpovey commented:
> Did you ever try p=1.0? I think in the original paper (https://arxiv.org/pdf/1904.08779v1.pdf) they never specified the probability of transform (`p` in the paper has - quite misleading - meaning of `mask_fraction` in this discussion), which to me implies they just applied the transform on all data.

I tried p=0.95,time_mask_fraction=0.15.  For some reason it was substantially worse than p=0.9, time_mask_fraction=0.15, see the table above.  I am not sure how much confidence I have in this result though.  I am trying p=0.85 as well.

luomingshuang commented:
I update some results:
![image](https://user-images.githubusercontent.com/37799481/154229570-84355f96-5b87-48a5-b49a-5043e63716dd.png)

luomingshuang commented:
Comparing with the baseline (original), I just modify the SpecAugment and change the value of max_mask_fraction and p.

danpovey commented:
You could also test this on tedlium, if you have enough GPUs.
I am worried that your experiments and my experiments are in disagreement about whether the
max_frames_mask_fraction should be 0.3 or 0.15. We need to explore that range, preferably
on other data.

luomingshuang commented:
OK. Tedlium3 will be trained over more quickly than the full librispeech.

luomingshuang commented:
The decoding results for some epochs:
![image](https://user-images.githubusercontent.com/37799481/155255524-84397de2-12ef-4369-a640-fbfd7deec676.png)

The above results show that the modified specaugment can accelerate the convergence of the model comparing to the original.
 
Some results based on 100h and full data:
![image](https://user-images.githubusercontent.com/37799481/155254455-367f9966-f0a7-4d0f-a279-79bdc45e4bc9.png)
![image](https://user-images.githubusercontent.com/37799481/155254475-0f23a720-0790-4e8d-8ee7-8e556e1330ae.png)

The best result trained with full data based on modified specaugment (max_mask_fraction=0.15, p=0.9) is `2.54 | 5.24`, while the public result is `2.42 | 5.73`.

In the other hand,  I also test it with tedlium3 transducer-stateless recipe. The results show as follows (not trained for many epochs):
![image](https://user-images.githubusercontent.com/37799481/155255996-95bb1971-f875-4339-a04b-6375bf11077c.png)

I think it's ok for us to use max-mask-fraction=0.15, p=0.9 as the fault value in the modified specaugment.

danpovey commented on egs/librispeech/ASR/tdnn_lstm_ctc/asr_datamodule.py:
```diff
@@ -383,3 +398,213 @@ def test_clean_cuts(self) -> CutSet:
+        :param num_frame_masks: how many frame (temporal) masks should be applied. Set to ``0`` to disable.
+        :param frames_mask_size: the width of the frame (temporal) masks (expressed in the number of masked frames).
+            This is the ``T`` parameter from the SpecAugment paper.
+        :param max_frames_mask_fraction: limits the size of the frame (temporal) mask to this value times the length
```
Please properly document the current meaning of this option, and of num-frame_masks!

danpovey commented:
Oh OK, yes, you seem to have a lot of experiments; I agree that (max_frames_mask_fraction=0.15, p=0.9) are probably reasonable defaults.

@pzelasko what do you think we should do on the Lhotse side?

danpovey commented:
BTW, if we want to publish about this, we'll also need the baseline results for p=1.0, which corresponds to the conventional SpecAug setting.

luomingshuang commented:
OK, will try this setting.
